dora/core/common/src/main/java/alluxio/membership/BarrierRecipe.java

```diff
@@ -0,0 +1,207 @@
+              new Cmp(key, Cmp.Op.GREATER, CmpTarget.createRevision(0L)))
+          .Then(Op.delete(key, DeleteOption.DEFAULT))
+          .Then(Op.put(key1, ByteSequence.EMPTY, PutOption.DEFAULT))
+          .commit();
+      TxnResponse txnResponse = txnResponseFut.get();
+      if (!txnResponse.isSucceeded()) {
+        throw new IOException("Failed to remove barrier for path:" + mBarrierPath);
+      }
+      LOG.info("Successfully remove barrier:{}", mBarrierPath);
+    } catch (ExecutionException | InterruptedException ex) {
+      LOG.error("Exception during removeBarrier.", ex);
+    }
+  }
+
+  /**
+   * Wait on barrier, waiting for the path to get deleted.
+   */
+  public void waitOnBarrierInternal() {
+    try {
+      Watch.Watcher watcher = mClient.getWatchClient().watch(
+          ByteSequence.EMPTY, WatchOption.newBuilder().build(), new Watch.Listener() {
+            @Override
+            public void onNext(WatchResponse response) {
+              WatchEvent event = response.getEvents().get(0);
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+              // NOOP
+            }
+
+            @Override
+            public void onCompleted() {
+              // NOOP
+            }
+          });
+      mClient.getWatchClient().watch(ByteSequence.from(mBarrierPath, StandardCharsets.UTF_8),
+          WatchOption.DEFAULT, watchResponse -> {
+            for (WatchEvent event : watchResponse.getEvents()) {
+              if (event.getEventType() == WatchEvent.EventType.DELETE
+                  && event.getKeyValue().getKey().equals(
+                      ByteSequence.from(mBarrierPath, StandardCharsets.UTF_8))) {
+                LOG.info("Delete event observed on path {}", mBarrierPath);
+                mLatch.countDown();
+              }
+            }
+          });
+      mLatch.await();
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+    LOG.info("Barrier wait done.");
+  }
+
+  /**
+   * Wait on barrier with no time restraint.
+   * @throws InterruptedException
+   */
+  public void waitOnBarrier() throws InterruptedException {
+    waitOnBarrierInternal();
+    mLatch.await();
+  }
+
+  /**
+   * Wait on barrier with a given timeout.
+   * @param time
+   * @param timeUnit
+   * @throws InterruptedException
+   */
+  @SuppressFBWarnings({"RV_RETURN_VALUE_IGNORED"})
+  public void waitOnBarrier(long time, TimeUnit timeUnit) throws InterruptedException {
+    waitOnBarrierInternal();
+    mLatch.await(time, timeUnit);
+  }
+
+  /**
+   * TEMPORARY simple barrier test - WIP.
+   * @param alluxioEtcdClient
+   */
+  public static void testBarrier(AlluxioEtcdClient alluxioEtcdClient) {
+    try {
+      BarrierRecipe barrierRecipe = new BarrierRecipe(alluxioEtcdClient, "/barrier-test",
+          "cluster1", 2L);
+      LOG.info("Setting barrier.");
+      barrierRecipe.setBarrier();
+      Thread t = new Thread(() -> {
```

Comment:
generally we are against using naked Thread, but this is a short testing command so it's no problem